specification/sbi_atee.adoc

```diff
@@ -938,6 +938,202 @@ The possible error codes returned in `sbiret.error` are shown below.
+| SBI_ERR_INVALID_PARAM   | `tvm_guest_id` or `base_page_address` or `length` were invalid.
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+[#sbi_tee_host_tvm_promote_page]
+=== Function: TEE Host TVM Promote Page (FID #18)
+[source, C]
+-----
+struct sbiret sbi_tee_host_tvm_promote_page(unsigned long tvm_guest_id,
+                                           unsigned long base_page_address,
+                                           unsigned long tsm_page_type);
+-----
+
+Promotes a set of contiguous mappings to the (one size up) requested page size.
+
+For each sub-page of the requested huge page mapping, the TSM verifies that:
+
+* The page is currently marked invalid in the TVM’s page table and has a size one smaller than
+the requested page size (e.g. 4kB if the requested size is 2MB).
+The pages are physically contiguous and suitably aligned to the requested huge page size.
+* The pages are all in the same state, either the `Blocked` state and uniquely owned by the TVM,
+or in the `BlockedShared` state and owned by the host. The TLB version at which the page was
+invalidated is older than the TVM’s current TLB version.
+
+After verifying these pre-conditions are met, the TSM:
+
+* Promotes the non-leaf PTE mapping the huge page to a leaf PTE with the PFN of the huge page and
+marks it as present.
+* Returns the now-unused page-table page to the TVM’s page-table page pool.
+* Places all the sub-pages in the `Mapped` state (if `Blocked`) or 'Shared' state (if `BlockedShared`).
+If the page was `BlockedShared`, the reference count of the page is restored.
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_tee_host_tvm_promote_page_errors]
+.TEE Host TVM Promote Page
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code              | Description
+| SBI_SUCCESS             | The operation completed successfully.
+| SBI_ERR_INVALID_PARAM   | `tvm_guest_id` or `base_page_address` or `tsm_page_type` were invalid.
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+[#sbi_tee_host_tvm_demote_page]
+=== Function: TEE Host TVM Demote Page (FID #19)
+[source, C]
+-----
+struct sbiret sbi_tee_host_tvm_demote_page(unsigned long tvm_guest_id,
+                                           unsigned long base_page_address,
+                                           unsigned long tsm_page_type);
+-----
+
+Demotes a huge page mapping to a set of contiguous mappings at the target size (one size down).
+
+The TSM verifies that:
+
+* The page is currently marked invalid in the TVM’s page table and has a size one larger than the
+requested page size (e.g. 2MB if the requested size is 4kB).
+* The page is in either the `Blocked` state and uniquely owned by the TVM, or in the `BlockedShared`
+state and owned by the host.
+* The TLB version at which the page was invalidated is older than the TVM’s current TLB version.
```

Comment:
same goes for the TLB version. The generic codes are too vague for these conditions.

Reply:
Commits removed. I will be opening a new PR for that after reworking the share/unshare parts.